### ISSUE TITLE
the meaning of weight_restrict_scaling_type?

### ISSUE THREAD

**Issue by YouanLi** [Xilinx/brevitas#212]
Hi,
I have a misunderstanding of this Attribute. I have learned in this [page](https://github.com/Xilinx/brevitas/issues/12), that the scaling factor is dividing by this weight_restrict_scaling_type, e.g. when weight_restrict_scaling_type is POWER_OF_TWO, the scaling factor should dividing by this POWER_OF_TWO number.

But when I learn about the weight distribution of weight_restrict_scaling_type, I find a misunderstanding. The following are the distributions of weight_restrict_scaling_type.POWER_OF_TWO and  weight_restrict_scaling_type.LOG_FP.
![image](https://user-images.githubusercontent.com/45391355/95430670-60bb8a80-094c-11eb-918a-e76a66a14510.png)
By comparing those two images, the scaling factor of POWER_OF_TWO which has a **smaller scaling factor**(because the original scaling factor should be dividing by POWER_OF_TWO). And the weight distribution of POWER_OF_TWO has over 1000 counts when x=0, which is more than LOG_FP. So I have a conclusion, the **smaller scaling factor** has more number of smaller **quantized weight values.** But this conclusion is violating the [quantization rule](https://github.com/Xilinx/brevitas/blob/49d5a1902c8f01521b344c0aae2d07ed672008ec/brevitas/core/quant.py#L655) . If the input is dividing by the scaling factor, the bigger the scaling factor the smaller the quantized weight.

My question is: 1. is my understanding of weight_restrict_scaling_type right?
                        2. is my understanding of the quantization rule right?

Best regards,

**Comment by volcacius** [comment]
Hello,

The scaling factor (let's call it s) is the smallest non zero value that you can represent with your quantization scheme. 
Let's say you quantize to 2 bits signed:

With s = 1 you a traditional integer representation, meaning you quantize any value to [-2, -1, 0, 1]. 
With s = 0.3, you have [-2 * 0.3, -1 * 0.3, 0 * 0.3, 1 * 0.3] = [-0.6, -0.3, 0, 0.3].
With s = 2 ^(-1) = 0.5, you have [1.0, 0.5, 0, 0.5]. This would be a Q1.1 fixed point.

restrict_scaling_type is a way to impose restriction on s. 
FP means no restriction, it can be an arbitrary floating point number. This is what is done also for quantization in Tensorflow, ONNX, etc.
LOG_FP still means no restriction, but it's _internally_ represented in log domain, which doesn't make a difference at inference time, but can make a difference at training time in terms of how the learning goes.
POWER_OF_TWO means that you want s to be a power of two, like in the example above with s = 2 ^ (-1), so that your overall quantization scheme maps to a Q format.

Hope this helps.

Alessandro

**Comment by YouanLi** [comment]
Hi Alessandro,

Thanks for your reply. And I have one more question about LOG_FP. Is this LOG_FP working as the following function: s=log(scaling factor). Is this LOG_FP an Attribute which designed by Logarithmic quantization?  Is there any code to describe this Attribute?

Best regards,

**Comment by volcacius** [comment]
Hello,

LOG_FP doesn't do any logarithmic quantization. 
It's designed for the following scenarios: let's say you want to learn s as a torch.nn.Parameter p, as I often do for activations (scaling_impl_type == PARAMETER). 
With FP, s = p. 
With LOG_FP, s = exp(p). So basically you are learning p as an exponent. 
Given the same learning rate, s move at different rates with FP or LOG_FP during training. 
In both cases, s is still an arbitrary floating point value.

Alessandro

**Comment by YouanLi** [comment]
Thanks for your quick reply and detailed explanation.
